File: python/cps_utility_console/menu.py
"""
CPS Utility Console — Menu Router
Author: Dominik Żurek
"""

from cps_utility_console import validators


def run_sensor_stats() -> None:
    """Run Sensor Statistics function"""
    from cps_utility_console.functions.sensor_stats import compute_stats

    print("\n--- Sensor Statistics ---")
    n = validators.get_positive_int("How many samples? ")

    samples = []
    for i in range(n):
        value = validators.get_float(f"Sample {i+1}/{n}: ")
        samples.append(value)

    result = compute_stats(samples)
    print("\nResults:")
    print(f"  Mean:   {result['mean']:.4f}")
    print(f"  Min:    {result['min']:.4f}")
    print(f"  Max:    {result['max']:.4f}")
    print(f"  StdDev: {result['stddev']:.4f}")


def run_unit_converter() -> None:
    """Run Unit Converter function."""
    from cps_utility_console.functions.unit_converter import (
        convert_temperature,
        convert_pressure,
    )
    
    print("\n--- Unit Converter ---")
    conv_type = validators.get_choice(
        "Convert [T]emperature or [P]ressure? ", ["T", "P"]
    )
    
    if conv_type == "T":
        value = validators.get_float("Enter temperature value: ")
        from_unit = validators.get_choice("From unit [C/F/K]: ", ["C", "F", "K"])
        to_unit = validators.get_choice("To unit [C/F/K]: ", ["C", "F", "K"])
        
        try:
            result = convert_temperature(value, from_unit, to_unit)
            print(f"\nResult: {value} {from_unit} = {result:.2f} {to_unit}")
        except ValueError as e:
            print(f"  Error: {e}")
    
    else:  # P
        value = validators.get_float("Enter pressure value: ")
        from_unit = validators.get_choice("From unit [BAR/KPA/PSI]: ", ["BAR", "KPA", "PSI"])
        to_unit = validators.get_choice("To unit [BAR/KPA/PSI]: ", ["BAR", "KPA", "PSI"])
        
        try:
            result = convert_pressure(value, from_unit, to_unit)
            print(f"\nResult: {value} {from_unit} = {result:.4f} {to_unit}")
        except ValueError as e:
            print(f"  Error: {e}")

def run_alarm_checker() -> None:
    """Placeholder for Alarm Threshold Checker function."""
    print(">> Alarm Threshold Checker — not implemented yet")


def run_lowpass_filter() -> None:
    """Placeholder for Low-Pass IIR Filter function."""
    print(">> Low-Pass IIR Filter — not implemented yet")


MENU_OPTIONS: dict[str, tuple[str, callable]] = {
    "1": ("Sensor Statistics",        run_sensor_stats),
    "2": ("Unit Converter",           run_unit_converter),
    "3": ("Alarm Threshold Checker",  run_alarm_checker),
    "4": ("Low-Pass IIR Filter",      run_lowpass_filter),
}


def display_menu() -> None:
    """Print the main menu to stdout."""
    print("\n=== CPS Utility Console ===")
    for key, (label, _) in MENU_OPTIONS.items():
        print(f"[{key}] {label}")
    print("[5] Exit")


def handle_choice(choice: str) -> bool:
    """
    Execute the selected menu option.
    Returns False if user wants to exit, True otherwise.
    """
    if choice == "5":
        print("Goodbye!")
        return False

    if choice in MENU_OPTIONS:
        _, func = MENU_OPTIONS[choice]
        func()
        return True

    print("  Error: invalid option. Please select 1-5.")
    return True

File: cpp/src/menu.cpp
/*
 * CPS Utility Console - Menu Router
 * Author: Dominik Żurek
 */

#include "../include/menu.hpp"
#include "../include/validators.hpp"
#include "../include/sensor_stats.hpp"
#include "../include/unit_converter.hpp"
#include "../include/alarm_checker.hpp"
#include "../include/lowpass_filter.hpp"
#include <iostream>
#include <iomanip>
#include <stdexcept>

void cps::runSensorStats() {
    std::cout << "\n--- Sensor Statistics ---\n";
    int n = cps::getPositiveInt("How many samples? ");

    std::vector<float> samples;
    for (int i = 0; i < n; i++) {
        float v = cps::getFloat("Sample " + std::to_string(i + 1) + "/" + std::to_string(n) + ": ");
        samples.push_back(v);
    }

    cps::SensorStats result = cps::computeStats(samples);
    std::cout << "\nResults:\n";
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "  Mean:   " << result.mean   << "\n";
    std::cout << "  Min:    " << result.min    << "\n";
    std::cout << "  Max:    " << result.max    << "\n";
    std::cout << "  StdDev: " << result.stddev << "\n";
}

void cps::runUnitConverter() {
    std::cout << "\n--- Unit Converter ---\n";
    std::string convType = cps::getChoice("Convert [T]emperature or [P]ressure? ", {"T", "P"});

    if (convType == "T") {
        float value = cps::getFloat("Enter temperature value: ");
        std::string fromUnit = cps::getChoice("From unit [C/F/K]: ", {"C", "F", "K"});
        std::string toUnit   = cps::getChoice("To unit [C/F/K]: ",   {"C", "F", "K"});
        try {
            float result = cps::convertTemperature(value, fromUnit, toUnit);
            std::cout << std::fixed << std::setprecision(2);
            std::cout << "\nResult: " << value << " " << fromUnit << " = " << result << " " << toUnit << "\n";
        } catch (const std::invalid_argument& e) {
            std::cout << "  Error: " << e.what() << "\n";
        }
    } else {
        float value = cps::getFloat("Enter pressure value: ");
        std::string fromUnit = cps::getChoice("From unit [BAR/KPA/PSI]: ", {"BAR", "KPA", "PSI"});
        std::string toUnit   = cps::getChoice("To unit [BAR/KPA/PSI]: ",   {"BAR", "KPA", "PSI"});
        try {
            float result = cps::convertPressure(value, fromUnit, toUnit);
            std::cout << std::fixed << std::setprecision(4);
            std::cout << "\nResult: " << value << " " << fromUnit << " = " << result << " " << toUnit << "\n";
        } catch (const std::invalid_argument& e) {
            std::cout << "  Error: " << e.what() << "\n";
        }
    }
}

void cps::runAlarmChecker() {
    std::cout << "\n--- Alarm Threshold Checker ---\n";
    float value = cps::getFloat("Enter sensor value: ");
    float low   = cps::getFloat("Enter LOW threshold: ");
    float high  = cps::getFloat("Enter HIGH threshold: ");

    try {
        std::string status = cps::checkAlarm(value, low, high);
        if (status == "OK")
            std::cout << "Status: " << status << ".\n";
        else if (status == "ALARM_LOW")
            std::cout << "Status: " << status << " (value below minimum).\n";
        else
            std::cout << "Status: " << status << " (value above maximum).\n";
    } catch (const std::invalid_argument& e) {
        std::cout << "  Error: " << e.what() << ".\n";
    }
}

void cps::runLowpassFilter() {
    std::cout << "\n--- Low-Pass IIR Filter ---\n";
    std::cout << "Formula: y[k] = a*x[k] + (1-a)*y[k-1]\n";

    int n     = cps::getPositiveInt("How many samples? ");
    float alpha = cps::getFloatInRange("Enter alpha [0.0-1.0]: ", 0.0f, 1.0f);

    std::vector<float> samples;
    for (int i = 0; i < n; i++) {
        float v = cps::getFloat("Sample " + std::to_string(i + 1) + "/" + std::to_string(n) + ": ");
        samples.push_back(v);
    }

    try {
        std::vector<float> filtered = cps::lowpassFilter(samples, alpha);
        std::cout << "\nResults:\n";
        std::cout << "Index | Input    | Filtered\n";
        std::cout << "------|----------|----------\n";
        std::cout << std::fixed << std::setprecision(4);
        for (int i = 0; i < (int)filtered.size(); i++) {
            std::cout << std::setw(5) << i
                      << " | " << std::setw(8) << std::setprecision(2) << samples[i]
                      << " | " << std::setw(8) << std::setprecision(4) << filtered[i] << "\n";
        }
    } catch (const std::invalid_argument& e) {
        std::cout << "  Error: " << e.what() << "\n";
    }
}

void cps::displayMenu() {
    std::cout << "\n=== CPS Utility Console ===\n";
    std::cout << "[1] Sensor Statistics\n";
    std::cout << "[2] Unit Converter\n";
    std::cout << "[3] Alarm Threshold Checker\n";
    std::cout << "[4] Low-Pass IIR Filter\n";
    std::cout << "[5] Exit\n";
}

bool cps::handleChoice(const std::string& choice) {
    if (choice == "5") { std::cout << "Goodbye!\n"; return false; }
    if (choice == "1") { cps::runSensorStats();    return true; }
    if (choice == "2") { cps::runUnitConverter();  return true; }
    if (choice == "3") { cps::runAlarmChecker();   return true; }
    if (choice == "4") { cps::runLowpassFilter();  return true; }
    std::cout << "  Error: invalid option. Please select 1-5.\n";
    return true;
}